2_process/src/lake_sequences_mntoha.py

```diff
@@ -0,0 +1,228 @@
+import numpy as np
+import pandas as pd
+
+def all_dates_depths(lake_obs_interpolated, 
+                     depths,
+                     temp_col='temp',
+                     depth_col='interpolated_depth',
+                     date_col='date',
+                     pad_before_days=0,
+                     pad_after_days=0,
+                     min_days=1):
+    """
+    Create a DataFrame with one column per depth and one row for every day from
+    `pad_before_days` days before the first observation until `pad_after_days`
+    after the last observation. Missing observations are filled in with np.NaN.
+
+    :param lake_obs_interpolated: DataFrame of observations for one lake, interpolated to the depths in the list `depths`.
+    :param depths: Full list of depths at which temperatures will be provided by the trained model.
+    :param temp_col: Name of DataFrame column with temperature values. (Default value = 'temp')
+    :param depth_col: Name of DataFrame column with depth values to match up with `depths`. (Default value = 'interpolated_depth')
+    :param date_col: Name of DataFrame column with observation dates. (Default value = 'date')
+    :param pad_before_days: Number of days to pad before the first observation (Default value = 0)
+    :param pad_after_days: Number of days to pad after the last observation (Default value = 0)
+    :param min_days: Minimum number of days included in output DataFrame (Default value = 1)
+    :returns: DataFrame of observations with NaN values for missing observations.
+
+    """
+
+    pad_before_days_td = np.timedelta64(pad_before_days, 'D')
+    pad_after_days_td = np.timedelta64(pad_after_days, 'D')
+    min_days_td = np.timedelta64(min_days, 'D')
+    one_day = np.timedelta64(1, 'D')
+    # start_date is pad_before_days before the date of the earliest observation
+    start_date = lake_obs_interpolated.date.min() - pad_before_days_td
+    # end_date is pad_after_days after the date of the latest observation
+    end_date = lake_obs_interpolated.date.max() + pad_after_days_td
+    # Ensure that at least min_days days are included
+    # If not, pad to the left (shift the start date earlier)
+    if end_date - start_date + one_day < min_days_td:
+        start_date = end_date - min_days_td + one_day
+    date_range = pd.date_range(start=start_date, end=end_date)
+
+    # Now take sparse observations specific to a few depths and a few days, 
+    # and create a full DataFrame of all depths and days, with NaN values for missing observations.
+
+    # Take only the columns we need
+    obs = lake_obs_interpolated[[date_col, depth_col, temp_col]]
```

Comment:
Oh I like this edit - it's much easier to follow and also makes your intentions plain